source/blog/2017-10-10-glimmer-progress-report.md

```diff
@@ -0,0 +1,645 @@
+   happen *before* the initial render starts.
+2. Internally, we need to order opcodes to align with HTML; that is, we need to
+   ensure that we always create an element and set its attributes before
+   attempting to append any child nodes. (Luckily for us, it happened to work
+   this way already.)
+
+We are hoping to test this approach in a real app in the next few weeks and
+report back on the results. We are tentatively optimistic that this will result
+in significant performance improvements, and suspect it may outperform even the
+best SSR implementations of Virtual DOM-based libraries like React and Preact
+because it requires fewer allocations.
+
+### Rehydration
+
+Rehydration is the ability of a client-side JavaScript app to "reconnect"
+components to the DOM generated by server-side rendered HTML. In many ways, a
+robust rehydration implementation is the holy grail that reunites the
+progressive enhancement and "all JavaScript all the time" camps.
+
+The server can serve up meaningful HTML, viewable even when JavaScript isn't
+available, and that HTML gets "progressively enhanced" with interactivity once
+the JavaScript loads and components rehydrate. But because components can _also_
+still produce their own DOM, you retain all of the benefits of client-side
+JavaScript, including the ability to work offline.
+
+Glimmer supports rehydration natively via its [ElementBuilder][element-builder] abstraction.
+In SSR mode, you can enable the [SerializeBuilder][serialize-builder], which includes additional
+comment annotations for where dynamic sections start and end.
+
+[element-builder]: https://github.com/glimmerjs/glimmer-vm/blob/72a1483/packages/@glimmer/runtime/lib/vm/element-builder.ts
+[serialize-builder]: https://github.com/glimmerjs/glimmer-vm/blob/72a1483/packages/%40glimmer/node/lib/serialize-builder.ts
+
+For example, given this template:
+
+```hbs
+<span class="user__name">{{@user.name}}</span>
+```
+
+The serialized output would include comments indicating the dynamic portion:
+
+```html
+<span class="user__name"><!--%+block:0%-->Chad Hietala<!--%-block:0%--></span>
+```
+
+In the browser, Glimmer is configured to use the
+[RehydrateBuilder][rehydrate-builder]. The RehydrateBuilder treats the existing
+DOM as a stack, and as the VM requests new elements get created, it compares the
+top of the "DOM stack" to the requested element. If it matches, the element is
+reused and the DOM is not mutated at all. In the case of a mismatch, the
+current block is cleared and the appropriate element is created and put in
+its place. The comment annotations are also stripped during this process, so you
+see a pristine DOM in your developer tools.
+
+[rehydrate-builder]: https://github.com/glimmerjs/glimmer-vm/blob/72a1483/packages/%40glimmer/runtime/lib/vm/rehydrate-builder.ts#L20
+
+We have rehydration working in our app, and intend to include it in the scope of
+the work to make SSR a first-class, easy-to-use API.
```

Comment:
This is a bit funky, I'm not sure what "scope of the work" means here. There isn't a milestone being described or anything to be scoped.

Reply:
"rehydration working experimentally in our app", maybe.